netutils/webclient/webclient.c

```diff
@@ -910,131 +1040,184 @@ int webclient_perform(FAR struct webclient_context *ctx)
-                                               ws->datend, &ws->buflen,
-                                               ctx->sink_callback_arg);
-                      if (ret != 0)
+                      /* Let the client decide what to do with the
+                       * received file.
+                       */
+
+                      if (ws->offset == ws->datend)
                         {
-                          goto errout_with_errno;
+                          /* We don't have data to give to the client yet. */
+                        }
+                      else if (ctx->sink_callback)
+                        {
+                          ret = ctx->sink_callback(&ws->buffer, ws->offset,
+                                                   ws->datend, &ws->buflen,
+                                                   ctx->sink_callback_arg);
+                          if (ret != 0)
+                            {
+                              goto errout_with_errno;
+                            }
+                        }
+                      else
+                        {
+                          ctx->callback(&ws->buffer, ws->offset, ws->datend,
+                                        &ws->buflen, ctx->sink_callback_arg);
                         }
                     }
                   else
                     {
-                      ctx->callback(&ws->buffer, ws->offset, ws->datend,
-                                    &ws->buflen, ctx->sink_callback_arg);
-                    }
-                }
-              else
-                {
-                  redirected = true;
-                  nredirect++;
-                  if (nredirect > CONFIG_WEBCLIENT_MAX_REDIRECT)
-                    {
-                      nerr("ERROR: too many redirects (%u)\n", nredirect);
-                      goto errout;
-                    }
+                      ws->nredirect++;
+                      if (ws->nredirect > CONFIG_WEBCLIENT_MAX_REDIRECT)
+                        {
+                          nerr("ERROR: too many redirects (%u)\n",
+                               ws->nredirect);
+                          ret = -ELOOP;
+                          goto errout_with_errno;
+                        }
 
-                  break;
+                      ws->state = WEBCLIENT_STATE_CLOSE;
+                      ws->redirected = 1;
+                      break;
+                    }
                 }
             }
         }
 
-      conn_close(ctx, &conn);
+      if (ws->state == WEBCLIENT_STATE_CLOSE)
+        {
+          conn_close(ctx, conn);
+          ws->need_conn_close = false;
+          if (ws->redirected)
+            {
+              ws->state = WEBCLIENT_STATE_SOCKET;
+            }
+          else
+            {
+              ws->state = WEBCLIENT_STATE_DONE;
+            }
+        }
     }
-  while (redirected);
+  while (ws->state != WEBCLIENT_STATE_DONE);
 
   free(ws);
```

Comment:
```suggestion
  free(ws);
  ctx->ws = NULL;
```
To avoid use-after-free errors, the freed pointer should be nullified.
```

Reply:
are you saying that there is a bug?
or is it just about a coding style?

Reply:
You may call it **defensive coding**.
Even after `free`, the `ctx->ws` pointer will still point to a memory address and since it is a member of a context structure it won't go out-of-scope. `ctx->ws` is now called a **dangling pointer**.
So every other check against `NULL` is simply useless if `ctx->ws` is not nullified, making the software prone to _user-after-free_ and _double-free_ kinds of errors.

Reply:
This is kinda pointless in situations like this. Dangling pointers are a thing if we are talking about global variables or library objects that are passed to library functions. But in this case? 1) return is called every time free() is called - thus ws is lost anyway 2) ws is accessed in this function directly via ws->, so even if you nullify this, you will still crash when calling ws->status or dereference ws in any way. Nullifying local variables before return is superflous.

Reply:
@mlyszczek 
my understanding is that his concern is about ctx->ws, not ws.

Reply:
@yamt
Ah yes, didn't look close enough, my bad. Yeah I think in this particular case nulling ctx->ws seems like a good idea. But I didn't check it good enough to be sure if that is necessary.